osu.Game/Rulesets/Mods/ModTimeRamp.cs

```diff
@@ -90,7 +88,7 @@ public double ApplyToRate(double time, double rate = 1)
 
         public virtual void Update(Playfield playfield)
         {
-            applyRateAdjustment(track.CurrentTime);
+            applyRateAdjustment(playfield.Clock.CurrentTime);
```

Comment:
Well, that's certainly no longer funny... definitely needs an inline comment mentioning that to avoid potentially changing it in the future, very spooky.

Not entirely sure how to solve this one, but one idea I can think of is providing the gameplay clock as part of `Playfield` and expose the source track clock from there.

Reply:
I personally wouldn't go there. You could change signature of `ApplyToTrack()` to something like

```csharp
void ApplyToTrack(IClock track, IAdjustableAudioComponent adjustments);
```

Or even just cast it once in `ModTimeRamp` and call it a day. I don't think exposing more stuff in the already-too-complex gameplay clock machinery serves anyone any good, but I'll wait and see what everyone else has to say about it.

Reply:
I would avoid storing a reference to yet another track in `Mod` at all costs IMO, and would want to completely decouple the idea of track consumption from `ApplyToTrack` which was only supposed to be for adding adjustments to the audio component at hand.

Reply:
I think that if exposed by `IClock` it's fine. You can't do anything bad with an `IClock`, unless you cast, at which point you've firmly blown your foot off on your own because that's not how interfaces work and cannot realistically complain that things are not working right.

Reply:
Maybe? For me at first, thinking about `ApplyToTrack` storing a track `IClock` reference inside and outside gameplay context yet being consumed in `Update(Playfield)` only kinda broke my mind there, as opposed to letting `Playfield` and `GameplayClock` expose the source time for the mod. Will wait for other opinions.

Reply:
I'm not entirely sure that the track clock is correct to use over the `FrameStableClock` specifically. Because in the end, it's that clock which determines the point of time which gameplay is at, the track clock is almost like a hint.

Don't we refer to the FrameStableClock in other mods?

Also thoughts @peppy?

Reply:
Thinking about it in terms of ramping is probably beneficial. Right now, when doing a seek, the audio seeks immediately (though I'd argue it shouldn't). If switching to providing the `FrameStableClock` here, the seek will be immediate but the ramp would be applied as the game catches up. It might feel a bit odd?

Reply:
I don't think this affects catchup? Because it's affecting the progression of time rather than the length of time, if that makes sense. Put another way, it affects how long a millisecond is _compared to real time_ but `T=1000` is always the same point in time no matter how it's reached.  
Catchup always happens at max ~80ms intervals (16ms frame stability * 5 update frames) independent of real-time.

Rather I was thinking about it from the perspective of offsets. I'm not sure if I'm correct since this is kinda hard to mentally parse, but if rate is applied relative to track time then it's possible that two systems with different offsets will have completely different playback.  
Suppose an arbitrary (either UO or local) offset of -1000ms meaning that game time is 1000ms behind track time (0ms track time = -1000ms gameplay time), and suppose that there's a hitobject at T=0. By the time gameplay hits T=0, the track time is going to be at T=1000ms and the approach circle for that hitobject will experience an increased time ramp compared to an offset of 0ms.

Reply:
For application of rate adjusts, I think applying to the "source of truth" is best. I think that's the track in this case. As long as we don't reverse that then the track's clock is best.

Correct me if I'm wrong, but the discussion here is proposing potentially reading the currrent time from `FrameStableClock` for use in
 https://github.com/ppy/osu/blob/8501a4161958514bfae52c5e78aee2ee26a870e0/osu.Game/Rulesets/Mods/ModTimeRamp.cs#L93

right?

If so, I think what I said is correct in that the rate applied will change during seeks.

Reply:
The frame stable clock _without_ offsets _might_ (and I'm saying this very carefully) be more correct than the track clock in one scenario specifically, which is seeking during replay playback. I've touched on that [in the past](https://github.com/ppy/osu/pull/11697#issuecomment-775733845), however I'm still pretty sure that any clock that takes user offsets into effect is not valid to use with `ModTimeRamp`.

Reply:
I’ve had thought about this for a bit and seeing what @smoogipoo has mentioned already, isn’t using `FrameStableClock` with offsets actually correct? 

What you want to be ramped is the gameplay clock, so that the speed starts changing as you progress through the hitobjects, constantly, regardless of any offset applied, which is what `GameplayClock/`FrameStableClock` should preserve.

When ramping with the source track clock, the speed starts change while you haven’t reached the first object yet, because the objects are offset from the track time.

Reply:
For gameplay, frame stable is the correct clock to use in all cases. I was focusing on the replay case because that's one area that will be less correct if changing this flow to use `FrameStableClock`.

But I think it's more correct to use it from a gameplay perspective, so I'd be in favour of that for now.

Reply:
Aside from the fact that now we're arguing on clock minutiae on a pull request supposing to fix an entirely unrelated regression, I acknowledge that using the track as a clock does potentially lead to a scenario wherein i.e. approach circles move slower or faster depending on offset, I still believe applying rate based on clocks with offset to be incorrect, because if you use that, then you actually ever so slightly modify the range of rate ramping (because at track time 0, frame stable clock w/offsets can be anywhere from -500 to +500ms). That seems more incorrect to me than visuals shifting.

That said, I'm mostly done with this conversation and I'll wait for everyone to make the final decision.

Reply:
To confirm we're on the same page: `FrameStableClock` has the offsets applied, so you are okay with using this, correct?

Reply:
I don't think that is correct, no, my claim remains that user offsets should not affect time ramp. But that's more or less a hunch based on the argument that the offset being applied is changing initial track rate, and the clock workings are so complex that I can't reasonably say whether I'm correct in this assertion.

Reply:
That means that two users with different user offsets would have different rates at the same hitobject, no? Seems wrong from a gameplay correctness/fairness perspective.

Reply:
I actually don't think either option is fair anymore. Either the visuals are sped up/slowed down depending on offset, or the track is. It feels more correct to me to leave the track in place as it's closer to being the "source" of time.

Reply:
@bdach Just to clarify, your case is when someone is abusing offsets, correct? I wasn't considering that case, only in the case where an offset is used to correct hitsound/gameplay <-> audio time, so for a non-malicious reason.

Reply:
Yeah as mentioned on discord seconds ago, if the offset application is to fix alignment with hitobjects, then I can agree that using post offset clock is correct.

Reply:
I'd much prefer this was a direct access to a `GameplayClock`, but I guess it's fine as-is.